plugins/ingest/src/main/java/org/elasticsearch/ingest/ValueSource.java

```
@@ -0,0 +1,196 @@
+ * under the License.
+ */
+
+package org.elasticsearch.ingest;
+
+import java.util.*;
+
+/**
+ * Holds a value. If the value is requested a copy is made and optionally template snippets are resolved too.
+ */
+// The best name I could come up with!
+public interface ValueSource {
+
+    /**
+     * Returns a copy of the value.
+     *
+     * For immutable values only a copy of the refence to the value is made.
+     * While copying template snippets are resolved based on the provided model.
+     */
+    Object copy(Map<String, Object> model);
+
+    static ValueSource wrap(Object value) {
+        return wrap(value, null);
+    }
+
+    static ValueSource wrap(Object value, TemplateService templateService) {
+        if (value instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<Object, Object> mapValue = (Map) value;
+            Map<ValueSource, ValueSource> valueTypeMap = new HashMap<>(mapValue.size());
+            for (Map.Entry<Object, Object> entry : mapValue.entrySet()) {
+                valueTypeMap.put(wrap(entry.getKey(), templateService), wrap(entry.getValue(), templateService));
+            }
+            return new MapValue(valueTypeMap);
+        } else if (value instanceof List) {
+            @SuppressWarnings("unchecked")
+            List<Object> listValue = (List) value;
+            List<ValueSource> valueSourceList = new ArrayList<>(listValue.size());
+            for (Object item : listValue) {
+                valueSourceList.add(wrap(item, templateService));
+            }
+            return new ListValue(valueSourceList);
+        } else if (value == null || value instanceof Integer ||
+            value instanceof Long || value instanceof Float ||
+            value instanceof Double || value instanceof Boolean) {
+            return new SimpleValue(value);
+        } else if (value instanceof String) {
+            if (templateService == null) {
+                return new SimpleValue(value);
```

are we sure we want to silently go ahead this way when templateService is null?